Helper/Discount.php

```diff
@@ -0,0 +1,352 @@
+            $quoteTable = $this->resource->getTableName('quote');
+
+            $sql = "DELETE FROM {$giftCardTable} WHERE quote_id IN 
+                    (SELECT entity_id FROM {$quoteTable} 
+                    WHERE bolt_parent_quote_id = :bolt_parent_quote_id AND entity_id != :entity_id)";
+
+            $connection->query($sql, ['entity_id' => $immutableQuoteId, 'bolt_parent_quote_id' => $parentQuoteId]);
+        } catch (\Zend_Db_Statement_Exception $e) {
+            $this->bugsnag->notifyException($e);
+        }
+    }
+
+    /**
+     * Remove Amasty Gift Card and update quote totals
+     *
+     * @param int $codeId
+     * @param Quote $quote
+     */
+    public function removeAmastyGiftCard($codeId, $quote)
+    {
+        if (! $this->isAmastyGiftCardAvailable()) {
+            return;
+        }
+        $connection = $this->resource->getConnection();
+        try {
+            $giftCardTable = $this->resource->getTableName('amasty_amgiftcard_quote');
+
+            $sql = "DELETE FROM {$giftCardTable} WHERE code_id = :code_id AND quote_id = :quote_id";
+            $connection->query($sql, ['code_id' => $codeId, 'quote_id' => $quote->getId()]);
+
+            $this->updateTotals($quote);
+        } catch (\Zend_Db_Statement_Exception $e) {
+            $this->bugsnag->notifyException($e);
+        }
+    }
+
+    /**
+     * Get Amasty "Pay for everything" (items, shipping and tax) additional configuration property.
+     * True by default if configuration is not explicitly set to false.
+     * The plugin currently supports both Amasty configurations,
+     * "Use a gift card to pay for shipping" and "Use a gift card to pay for tax" set to either "Yes" or "No".
+     * Any "Yes" + "No" combination is not supported by Bolt.
+     *
+     * @return bool
+     */
+    public function getAmastyPayForEverything()
+    {
+        $amastyGiftCardConfig = $this->configHelper->getAmastyGiftCardConfig();
+        if ($amastyGiftCardConfig && @$amastyGiftCardConfig->payForEverything === false) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Get Amasty Gift Card codes, stored comma separated in total title field, return as array.
+     *
+     * @param array $totals totals array collected from quote
+     * @return array
+     */
+    public function getAmastyGiftCardCodesFromTotals($totals) {
+        $giftCardCodes = explode(',', $totals[self::AMASTY_GIFTCARD]->getTitle());
+        return array_map('trim', $giftCardCodes);
+    }
+
+    /**
+     * Get accumulated unused balance of all Amasty Gift Cards corresponding to passed gift card coupons array
+     *
+     * @param array $giftCardCodes
+     * @return float|int|mixed
+     */
+    public function getAmastyGiftCardCodesCurrentValue($giftCardCodes) {
+        $amount = 0;
+        foreach ($giftCardCodes as $giftCardCode) {
+            $giftCard = $this->loadAmastyGiftCard($giftCardCode);
```

Comment:
This seems to be running a lot of queries to get each gift card value. Would something like this work better?
`$giftCardCollection->addFieldToFilter('current_value', ['in'=>$giftCardCodes])->getData();`
And then summarize the current_value on the returned array.

Reply:
Thanks @johnmccombs1. Updated.